agent/src/agent/data_extractor/cacti/repository.py

```diff
@@ -10,98 +10,136 @@ def _get_session(connection_string: str):
-
-        variables[local_graph_id][row['field_name']] = row['field_value']
-    return variables
-
-
-def get_hosts(mysql_connection_string: str) -> dict:
-    hosts = {}
-    res = _get_session(mysql_connection_string).execute("""
-        SELECT description, hostname, snmp_community, snmp_version, snmp_username, snmp_password, snmp_auth_protocol,
-         snmp_priv_passphrase, snmp_context, snmp_port, snmp_timeout, ping_retries, max_oids, id
-        FROM host
-    """)
-    for row in res:
-        hosts[row['id']] = dict(row)
-    return hosts
-
-
-def get_graphs(mysql_connection_string: str) -> dict:
-    graphs = {}
-    res = _get_session(mysql_connection_string).execute("""
-        SELECT local_graph_id, title FROM graph_templates_graph
-    """)
-    for row in res:
-        graphs[row['local_graph_id']] = row['title']
-    return graphs
-
-
-def get_sources(mysql_connection_string: str) -> dict:
-    data_sources = {}
-    res = _get_session(mysql_connection_string).execute("""
-        SELECT DISTINCT gti.local_graph_id, dtd.data_source_path
-        FROM graph_templates_item AS gti
-        JOIN data_template_rrd AS dtr ON gti.task_item_id = dtr.id
-        JOIN data_template_data dtd on dtd.local_data_id = dtr.local_data_id
-    """)
-    for row in res:
-        data_sources[row['local_graph_id']] = row['data_source_path']
-    return data_sources
+            JOIN data_template_data dtd on dtd.local_data_id = dtr.local_data_id
+            WHERE local_graph_id != 0
+            AND data_source_path IS NOT NULL
+            AND gti.graph_type_id IN (4, 5, 6, 7 ,8);
+        """)
+        for row in res:
+            local_graph_id = row['local_graph_id']
+            if local_graph_id not in self.graphs:
+                self.graphs[local_graph_id] = {}
+            if 'items' not in self.graphs[local_graph_id]:
+                self.graphs[local_graph_id]['items'] = {}
+            self.graphs[local_graph_id]['items'][row['item_id']] = {'data_source_path': row['data_source_path']}
+
+    def _get_graph_titles(self):
+        res = self.session.execute("""
+            SELECT gtg.local_graph_id, gtg.title, gl.host_id
+            FROM graph_templates_graph gtg
+            JOIN graph_local gl ON gtg.local_graph_id = gl.id
+            WHERE local_graph_id != 0
+        """)
+        for row in res:
+            if row['local_graph_id'] not in self.graphs:
+                continue
+            self.graphs[row['local_graph_id']]['title'] = row['title']
+            self.graphs[row['local_graph_id']]['host_id'] = str(row['host_id'])
+
+    def _get_graph_variables(self):
+        res = self.session.execute("""
+            select gl.id, field_name, field_value
+            from host_snmp_cache hsc
+            join graph_local gl
+            on gl.host_id = hsc.host_id
+            and gl.snmp_index = hsc.snmp_index
+            and gl.snmp_query_id = hsc.snmp_query_id
+        """)
```

Comment:
How long does this query execute? Maybe we need to force index here also

Reply:
I didn't notice any problems with this query, when I run it in the console it takes about 3 seconds

Reply:
If you mean console in pycharm then it selects only 500 rows this is why it works fast there. If you check with `count` statement it takes 50 seconds. With forcing index `count` takes 3,5 seconds. So I think you should add it